### PR TITLE
825 change string comparisons to constant time comparisons in auth middleware library

### LINKED ISSUE #825: [SECURITY] Change string comparisons to constant time comparisons in auth middleware library
**Describe the solution you'd like**
We need to change string comparisons in auth middleware lib (especially the ones that compare tokens, keys, codes, etc.). We can use hmac.compare_digest.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ywkim312**
## Description
Changed string comparison to constant time comparison in auth middleware library.

**Fixes #825**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [x] Within a week
- [ ] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

File: contributions/catalog/requirements.txt [hunk 1]
```diff
@@ -4,4 +4,7 @@ gunicorn==19.9.0
 requests==2.22.0
 requests-oauthlib==1.3.0
 gevent==20.9.0
+<<<<<<< HEAD
```
Comment (edthedev) [inline contributions/catalog/requirements.txt:7]: @ywkim312 Looks like a merge error happened here.

Comment (ywkim312) [reply contributions/catalog/requirements.txt]: fixed the merge conflict and changed the merge base from develop to master

**Comment by sandeep-ps** [comment]
Hi, @ywkim312, 

I understand that initially, I had suggested creating a PR against the master branch for this task. Since we have recently completed a release and submitted the deployment requests, this can go in the next release. Could you please change the destination to develop branch? Sorry about that.

**Comment by ywkim312** [comment]
> Hi, @ywkim312,
> 
> I understand that initially, I had suggested creating a PR against the master branch for this task. Since we have recently completed a release and submitted the deployment requests, this can go in the next release. Could you please change the destination to develop branch? Sorry about that.

No worries, changed to develop

File: Dockerfile [hunk 2]
```diff
@@ -20,7 +20,7 @@ COPY eventservice/events.yaml /usr/share/nginx/html/app/
 COPY profileservice/profile.yaml /usr/share/nginx/html/app/
 COPY loggingservice/logging.yaml /usr/share/nginx/html/app/
 
-ENV URLS "[{url: 'app/appconfig.yaml', name: 'App Config Building Block'}, {url: 'app/auth.yaml', name: 'Authentication Building Block'}, {url: 'app/events.yaml', name: 'Events Building Block'}, {url: 'app/profile.yaml', name: 'Profile Building Block'}, {url: 'app/logging.yaml', name: 'Logging Building Block'}, {url: 'https://api-dev.rokwire.illinois.edu/health/doc', name: 'Health Building Block'}, {url: 'https://api-dev.rokwire.illinois.edu/talent-chooser/doc', name: 'Talent Chooser Building Block'} ]"
+ENV URLS "[{url: 'app/appconfig.yaml', name: 'App Config Building Block'}, {url: 'app/auth.yaml', name: 'Authentication Building Block'}, {url: 'app/events.yaml', name: 'Events Building Block'}, {url: 'app/profile.yaml', name: 'Profile Building Block'}, {url: 'app/logging.yaml', name: 'Logging Building Block'}, {url: 'https://api.rokwire.illinois.edu/health/doc', name: 'Health Building Block'}, {url: 'https://api.rokwire.illinois.edu/talent-chooser/doc', name: 'Talent Chooser Building Block'} ]"
```
Comment (sandeep-ps) [inline Dockerfile:23]:
```suggestion
ENV URLS "[{url: 'app/appconfig.yaml', name: 'App Config Building Block'}, {url: 'app/auth.yaml', name: 'Authentication Building Block'}, {url: 'app/events.yaml', name: 'Events Building Block'}, {url: 'app/profile.yaml', name: 'Profile Building Block'}, {url: 'app/logging.yaml', name: 'Logging Building Block'}, {url: 'https://api-dev.rokwire.illinois.edu/health/doc', name: 'Health Building Block'}, {url: 'https://api-dev.rokwire.illinois.edu/talent-chooser/doc', name: 'Talent Chooser Building Block'} ]"
```

Suggesting this change as probably the source branch was master.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,8 +4,9 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased]
+## [unreleased]
```
Comment (sandeep-ps) [inline CHANGELOG.md:7]:
```suggestion
## [Unreleased] 
```

Comment (ywkim312) [reply CHANGELOG.md:7]: changed

File: Dockerfile (hunk 2 above)
Comment (ywkim312) [reply Dockerfile:23]: changed

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### Changed
+- String comparisons to constant time comparisons in auth middleware library. [#825](https://github.com/rokwire/rokwire-building-blocks-api/issues/825)
```
Comment (sandeep-ps) [inline CHANGELOG.md:9]: This needs to be moved under the Security section. Please see https://keepachangelog.com/en/1.0.0/#how. Thanks.

**Comment by sandeep-ps** [comment]
The Postman Tests look fine. I will approve and merge after the CHANGELOG update. Thanks.

File: CHANGELOG.md (hunk 4 above)
Comment (ywkim312) [reply CHANGELOG.md]: didn't know about that. changed.

**Comment by sandeep-ps** [comment]
I have fixed the merge conflict that I introduced by merging the other PR. :-D

**Comment by ywkim312** [comment]
> I have fixed the merge conflict that I introduced by merging the other PR. :-D

haha thanks

**Comment by sandeep-ps** [comment]
@edthedev, Do you have any other comments? If not, I was thinking of merging this soon. Thanks for your feedback on this and the other PR.

**Review by edthedev — APPROVED** [review]

**Comment by edthedev** [comment]
@sandeep-ps Looks good. Thanks!

**Review by sandeep-ps — APPROVED** [review]

**Pull request merged**